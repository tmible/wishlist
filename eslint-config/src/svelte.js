import svelte from "eslint-plugin-svelte";
import globals from "globals";

export default [
  ...svelte.configs["flat/recommended"],
  {
    languageOptions: {
      globals: {
        ...globals.browser,
        ...globals.node,
      },
      ecmaVersion: "latest",
      sourceType: "module",
      parserOptions: {
        extraFileExtensions: [ ".svelte" ],
      },
    },

    settings: {
      "import/parsers": {
        "svelte-eslint-parser": [ ".svelte" ],
        "espree": [ ".js" ],
      },
    },

    rules: {
      /* import */
      // svelte does tree shaking when compiling
      "import/no-namespace": "off",

      /* svelte */
      "svelte/infinite-reactive-loop": "error",
      "svelte/no-dom-manipulating": "error",
      "svelte/no-dupe-on-directives": "error",
      "svelte/no-dupe-use-directives": "error",
      "svelte/no-reactive-reassign": "error",
      "svelte/no-store-async": "error",
      "svelte/require-store-callbacks-use-set-param": "error",
      "svelte/block-lang": "error",
      "svelte/no-ignored-unsubscribe": "error",
      "svelte/no-immutable-reactive-statements": "error",
      "svelte/no-inline-styles": "error",
      "svelte/no-reactive-functions": "error",
      "svelte/no-reactive-literals": "error",
      "svelte/no-useless-mustaches": "error",
      "svelte/require-each-key": "error",
      "svelte/require-event-dispatcher-types": "error",
      "svelte/require-stores-init": "error",
      "svelte/valid-each-key": "error",
      "svelte/derived-has-same-inputs-outputs": "error",
      "svelte/first-attribute-linebreak": "error",
      "svelte/html-closing-bracket-new-line": "error",
      "svelte/html-closing-bracket-spacing": "error",
      "svelte/html-quotes": "error",
      "svelte/html-self-closing": [
        "error",
        {
          "void": "never",
          "normal": "never",
          "svg": "always",
          "math": "never",
          "component": "always",
          "svelte": "always",
        },
      ],
      "svelte/indent": [ "error", { ignoredNodes: [ "ConditionalExpression" ] } ],
      "svelte/max-attributes-per-line": [ "error", { multiline: 1, singleline: 100 } ],
      "svelte/mustache-spacing": "error",
      "svelte/no-extra-reactive-curlies": "error",
      "svelte/no-spaces-around-equal-signs-in-attribute": "error",
      "svelte/prefer-class-directive": "error",
      "svelte/prefer-style-directive": "error",
      "svelte/shorthand-attribute": "error",
      "svelte/sort-attributes": [
        "error",
        {
          order: [
            // `this` property.
            "this",
            // `bind:this` directive.
            "bind:this",
            // `id` attribute.
            "id",
            // `name` attribute.
            "name",
            // `slot` attribute.
            "slot",
            // `--style-props`
            { match: "/^--/u", sort: "ignore" },
            // `style` attribute, and `style:` directives.
            [ "style", "/^style:/u" ],
            // `class` attribute.
            "class",
            // `class:` directives.
            { match: "/^class:/u", sort: "ignore" },
            // other attributes.
            {
              match: [ "!/:/u", "!/^(?:this|id|name|style|class)$/u", "!/^--/u" ],
              sort: "ignore",
            },
            // `bind:` directives (other then `bind:this`), and `on:` directives.
            [ "/^bind:/u", "!bind:this", "/^on:/u" ],
            // `use:` directives.
            { match: "/^use:/u", sort: "ignore" },
            // `transition:` directive.
            { match: "/^transition:/u", sort: "ignore" },
            // `in:` directive.
            { match: "/^in:/u", sort: "ignore" },
            // `out:` directive.
            { match: "/^out:/u", sort: "ignore" },
            // `animate:` directive.
            { match: "/^animate:/u", sort: "ignore" },
            // `let:` directives.
            { match: "/^let:/u", sort: "ignore" },
          ],
        },
      ],
      "svelte/spaced-html-comment": "error",
      "svelte/no-inner-declarations": "error",
      "svelte/no-trailing-spaces": "error",
    },
  },
  {
    files: [ "**/*.svelte" ],
    rules: {
      // svelte components can import themselves
      "import/no-self-import": "off",
    },
  },
  {
    files: [ "**/**.test.js", "**/**.test.svelte.js", "**/__tests__/**" ],
    settings: { "import/ignore": [ "vitest", "@testing-library/svelte" ] },
  },
];
