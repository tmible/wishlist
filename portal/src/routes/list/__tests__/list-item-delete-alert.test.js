// @vitest-environment jsdom
import { cleanup, render, screen } from '@testing-library/svelte';
import { userEvent } from '@testing-library/user-event';
import { afterEach, beforeAll, beforeEach, describe, expect, it, vi } from 'vitest';

describe('list item delete alert', () => {
  let dispatchSpies;

  beforeAll(() => {
    dispatchSpies = [];
    vi.doMock(
      'svelte',
      async (importOriginal) => {
        const original = await importOriginal();
        return {
          ...original,
          createEventDispatcher: () => {
            const spy = vi.fn(original.createEventDispatcher());
            dispatchSpies.push(spy);
            return spy;
          },
        };
      },
    );
  });

  afterEach(() => {
    dispatchSpies = [];
    vi.clearAllMocks();
    cleanup();
  });

  it('should be displayed closed', async () => {
    const { baseElement } = render(
      await import('../list-item-delete-alert.svelte').then((module) => module.default),
      { open: false, listItemToDelete: { id: 'id', name: 'name' } },
    );
    expect(baseElement).toMatchSnapshot();
  }, 10000);

  describe('opened', () => {
    let baseElement;

    beforeEach(async () => {
      vi.stubGlobal('fetch', vi.fn());
      ({ baseElement } = render(
        await import('../list-item-delete-alert.svelte').then((module) => module.default),
        { open: true, listItemToDelete: { id: 'id', name: 'name' } },
      ));
    });

    it('should be displayed', () => {
      // remove random ids to match snapshot
      const alertDialog = screen.getByRole('alertdialog');
      alertDialog.removeAttribute('id');
      const titleId = alertDialog.getAttribute('aria-describedby');
      alertDialog.removeAttribute('aria-describedby');
      const descriptionId = alertDialog.getAttribute('aria-labelledby');
      alertDialog.removeAttribute('aria-labelledby');
      baseElement.querySelector(`[id="${titleId}"]`).removeAttribute('id');
      baseElement.querySelector(`[id="${descriptionId}"]`).removeAttribute('id');
      expect(baseElement).toMatchSnapshot();
    });

    describe('on cancel button click', () => {
      beforeEach(async () => {
        const user = userEvent.setup();
        await user.click(screen.getByText('Отмена', { selector: 'button' }));
      });

      it('should close', () => {
        expect(baseElement).toMatchSnapshot();
      });

      it('should not request deletion', () => {
        expect(vi.mocked(fetch)).not.toHaveBeenCalled();
      });

      it('should not dispatch event', () => {
        expect(dispatchSpies[0]).not.toHaveBeenCalled();
      });
    });

    describe('on confirm button click', () => {
      beforeEach(async () => {
        const user = userEvent.setup();
        await user.click(screen.getByText('Удалить', { selector: 'button' }));
      });

      it('should close', () => {
        expect(baseElement).toMatchSnapshot();
      });

      it('should request deletion', () => {
        expect(
          vi.mocked(fetch),
        ).toHaveBeenCalledWith(
          '/api/wishlist',
          {
            method: 'DELETE',
            headers: { 'Content-Type': 'application/json;charset=utf-8' },
            body: JSON.stringify([ 'id' ]),
          },
        );
      });

      it('should dispatch event', () => {
        expect(dispatchSpies[0]).toHaveBeenCalledWith('delete');
      });
    });
  });
});
