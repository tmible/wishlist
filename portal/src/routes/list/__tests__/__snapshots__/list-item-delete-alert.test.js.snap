// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`list item delete alert > opened > on cancel button click > should close 1`] = `
<body
  style=""
>
  <div>
    <button
      aria-expanded="false"
      aria-haspopup="dialog"
      class="hidden"
      data-alert-dialog-trigger=""
      data-melt-dialog-trigger=""
      type="button"
    />
    
     
    
  </div>
  <div
    data-alert-dialog-portal=""
    data-melt-dialog-portalled=""
    data-portal=""
  >
    
     
    
  </div>
</body>
`;

exports[`list item delete alert > opened > on confirm button click > should close 1`] = `
<body
  style=""
>
  <div>
    <button
      aria-expanded="false"
      aria-haspopup="dialog"
      class="hidden"
      data-alert-dialog-trigger=""
      data-melt-dialog-trigger=""
      type="button"
    />
    
     
    
  </div>
  <div
    data-alert-dialog-portal=""
    data-melt-dialog-portalled=""
    data-portal=""
  >
    
     
    
  </div>
</body>
`;

exports[`list item delete alert > opened > should be displayed 1`] = `
<body
  data-melt-scroll-lock=""
  style="overflow: hidden; padding-right: calc(+ 1024px);"
>
  <div>
    <button
      aria-expanded="true"
      aria-haspopup="dialog"
      class="hidden"
      data-alert-dialog-trigger=""
      data-melt-dialog-trigger=""
      type="button"
    />
    
     
    
  </div>
  <div
    data-alert-dialog-portal=""
    data-melt-dialog-portalled=""
    data-portal=""
  >
    <div
      aria-hidden="true"
      class="modal-overlay"
      data-alert-dialog-overlay=""
      data-escapee=""
      data-melt-dialog-overlay=""
      data-state="open"
      style=""
      tabindex="-1"
    />
    
     
    <div
      aria-modal="true"
      class="modal-box modal-alert"
      data-alert-dialog-content=""
      data-escapee=""
      data-melt-dialog-content=""
      data-state="open"
      role="alertdialog"
      style=""
      tabindex="-1"
    >
      <div
        class="contents prose"
      >
        <h2
          data-alert-dialog-title=""
          data-melt-dialog-title=""
        >
          Подвердите удаление
        </h2>
        
        
         
        <p
          data-alert-dialog-description=""
          data-melt-dialog-description=""
        >
          Вы уверены, что хотите удалить желание «
          name
          » из списка?
        </p>
        
         
        <div
          class="card-actions"
        >
          <button
            class="btn btn-neutral flex-1"
            data-alert-dialog-cancel=""
            data-melt-dialog-close=""
            type="button"
          >
            Отмена
          </button>
          
           
          <button
            class="btn btn-error flex-1"
            data-alert-dialog-action=""
            data-alert-dialog-cancel=""
            data-melt-dialog-close=""
            type="button"
          >
            Удалить
          </button>
          
        </div>
      </div>
    </div>
    
  </div>
</body>
`;

exports[`list item delete alert > should be displayed closed 1`] = `
<body>
  <div>
    <button
      aria-expanded="false"
      aria-haspopup="dialog"
      class="hidden"
      data-alert-dialog-trigger=""
      data-melt-dialog-trigger=""
      type="button"
    />
    
     
    
  </div>
  <div
    data-alert-dialog-portal=""
    data-melt-dialog-portalled=""
    data-portal=""
  >
    
     
    
  </div>
</body>
`;
